File: backend/src/main/java/com/demo/travel_expense_management/travelrequest/mapper/TravelRequestMapper.java
package com.demo.travel_expense_management.travelrequest.mapper;

import java.util.List;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;

import com.demo.travel_expense_management.travelrequest.dto.request.CreateTravelRequestRequest;
import com.demo.travel_expense_management.travelrequest.dto.request.UpdateTravelRequestRequest;
import com.demo.travel_expense_management.travelrequest.dto.response.TravelRequestResponse;
import com.demo.travel_expense_management.travelrequest.dto.response.TravelRequestSummaryResponse;
import com.demo.travel_expense_management.travelrequest.entity.TravelRequest;

@Mapper(componentModel = "spring")
public interface TravelRequestMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "requestNumber", ignore = true)
    @Mapping(target = "applicant", ignore = true)
    @Mapping(target = "status", ignore = true)
    @Mapping(target = "observations", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "purpose", source = "travelPurpose")
    TravelRequest toEntity(CreateTravelRequestRequest request);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "requestNumber", ignore = true)
    @Mapping(target = "applicant", ignore = true)
    @Mapping(target = "status", ignore = true)
    @Mapping(target = "observations", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "purpose", source = "travelPurpose")
    void updateEntity(UpdateTravelRequestRequest request, @MappingTarget TravelRequest travelRequest);

    @Mapping(target = "approvedAmount", ignore = true)
    @Mapping(target = "applicantId", source = "applicant.id")
    @Mapping(target = "applicantFullName", expression = "java(travelRequest.getApplicant().getFirstName() + \" \" + travelRequest.getApplicant().getLastName())")
    @Mapping(target = "departmentId", source = "applicant.department.id")
    @Mapping(target = "departmentName", source = "applicant.department.name")
    @Mapping(target = "travelPurpose", source = "purpose")
    TravelRequestResponse toResponse(TravelRequest travelRequest);

    @Mapping(target = "applicantFullName", expression = "java(travelRequest.getApplicant().getFirstName() + \" \" + travelRequest.getApplicant().getLastName())")
    TravelRequestSummaryResponse toSummaryResponse(TravelRequest travelRequest);

    List<TravelRequestSummaryResponse> toSummaryResponseList(List<TravelRequest> travelRequests);
}

File: frontend/src/lib/mapper/travel-request.mapper.ts
import type {
	CreateTravelRequestDto,
	TravelRequestResponseDto,
	TravelRequestSummaryResponseDto,
} from '$lib/dto/travel-request';
import type {
	TravelRequest,
	TravelRequestFormValues,
	TravelRequestSummary,
} from '$lib/models/travel-request';

export function toCreateTravelRequestDto(
	values: TravelRequestFormValues,
): CreateTravelRequestDto {
	return {
		travelPurpose: values.travelPurpose.trim(),
		destination: values.destination.trim(),
		departureDate: values.departureDate.trim(),
		returnDate: values.returnDate.trim(),
		requestedAmount: Number(values.requestedAmount),
	};
}

export function toTravelRequestSummary(
	dto: TravelRequestSummaryResponseDto,
): TravelRequestSummary {
	return {
		id: dto.id,
		requestNumber: dto.requestNumber,
		applicantFullName: dto.applicantFullName,
		destination: dto.destination,
		requestedAmount: dto.requestedAmount,
		status: dto.status,
		departureDate: dto.departureDate,
		returnDate: dto.returnDate,
	};
}

export function toTravelRequest(dto: TravelRequestResponseDto): TravelRequest {
	return {
		id: dto.id,
		requestNumber: dto.requestNumber,
		applicantId: dto.applicantId,
		applicantFullName: dto.applicantFullName,
		departmentId: dto.departmentId,
		departmentName: dto.departmentName,
		travelPurpose: dto.travelPurpose,
		destination: dto.destination,
		departureDate: dto.departureDate,
		returnDate: dto.returnDate,
		requestedAmount: dto.requestedAmount,
		approvedAmount: dto.approvedAmount,
		status: dto.status,
		createdAt: dto.createdAt,
		updatedAt: dto.updatedAt,
	};
}
